From d57efb0fb6b7f433faaafe3339e9c87a4abc7605 Mon Sep 17 00:00:00 2001
From: Catalin Udma <catalin-dan.udma@nxp.com>
Date: Fri, 2 Apr 2021 18:38:55 +0300
Subject: [PATCH 39/42] pfe: fixup dma-coherent to support both Rev1 and Rev2

Rev2 requires dma-coherent - this is not a silicon requirement but a
driver performance requirement for Rev2.
If U-Boot detects Rev1 then fixup Linux's device tree and remove
dma-coherent

Upstream-Status: Pending

Issue: ALB-6842

Signed-off-by: Catalin Udma <catalin-dan.udma@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 board/freescale/s32-gen1/eth.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/board/freescale/s32-gen1/eth.c b/board/freescale/s32-gen1/eth.c
index c430c950a6..3372488911 100644
--- a/board/freescale/s32-gen1/eth.c
+++ b/board/freescale/s32-gen1/eth.c
@@ -125,6 +125,14 @@ static void ft_enet_pfe_emac_fixup(u32 idx, void *fdt)
 	}
 }
 
+static void ft_enet_coherent_fixup(void *fdt, int nodeoff)
+{
+	if (is_s32gen1_soc_rev1())
+		if (fdt_getprop(fdt, nodeoff, "dma-coherent", NULL))
+			if (fdt_delprop(fdt, nodeoff, "dma-coherent"))
+				pr_err("Failed to remove dma-coherent\n");
+}
+
 static bool pfeng_drv_status_active(void)
 {
 	struct udevice *dev;
@@ -161,6 +169,9 @@ void ft_enet_fixup(void *fdt)
 			ft_enet_pfe_emac_fixup(0, fdt);
 			ft_enet_pfe_emac_fixup(1, fdt);
 			ft_enet_pfe_emac_fixup(2, fdt);
+
+			/* Remove dma-coherent for Rev 1*/
+			ft_enet_coherent_fixup(fdt, nodeoff);
 		}
 	}
 #endif /* CONFIG_IS_ENABLED(FSL_PFENG) */
-- 
2.25.1

